design/runnable.go

```diff
@@ -78,125 +96,581 @@ var _ = Service("runnable", func() {
+ the location where the artifact's contents are uploaded. 
+
+Depending on the concrete provider type decided at runtime, the expression {{ output.<output-name> }}
+will be expanded differently:
+- local: the expression is resolved to the local container path where contents are expected by
+the framework
+- inline: expression is expanded to the actual artifact contents
+- fuseml: the FuseML codeset resource URL
+- everything else: the remote URL where the artifact contents are stored
+The expression {{ output.<output-name>.path }} will be expanded to the container path where the
+framework expects the artifact contents or URL value to be provided by the runnable implementation.
+`, func() {
+			Example([]string{
+				"local",
+			})
+			Example([]string{
+				"fuseml",
+				"s3",
+			})
+			Example([]string{
+				"git",
+				"http",
+				"https",
+				"ftp",
+				"sftp",
+			})
+			Default([]string{"local"})
+		})
+	Field(4, "kind", RunnableOutputArtifactKind,
+		`The kind of output artifact and its specific attributes. For general resources and artifacts, this field is not set.
+These attributes describe the contents of artifacts generated by the runnable as output.`)
+	Field(5, "path", String,
+		"Specify a custom container path where the artifact contents or the artifact URL(s) are provided by the container", func() {
+			Example("/workspace/output/models")
+		})
+	Field(7, "dimension", String,
+		"Number of artifacts encoded by this output. Use 'single' to encode a single artifact and 'array' to encode multiple",
+		func() {
+			Enum("single", "array")
+			Example("single")
+			Default("single")
+		})
+	Field(8, "labels", MapOf(String, String),
+		`List of multi-purpose labels. Used to further label the artifacts that are generated as output by this runnable.`,
+		func() {
+			Key(func() {
+				Pattern(identifierPattern)
+			})
+			Example(map[string]string{
+				"library":  "pytorch",
+				"function": "predict",
+			})
+		})
+})
+
+// RunnableInputArtifactKind encodes the type of input artifact as well as type-specific attributes
+var RunnableInputArtifactKind = Type("RunnableInputArtifactKind", func() {
+	Field(1, "codeset", CodesetArgumentDesc,
+		"Input codeset attributes.")
+	Field(2, "model", ModelArgumentDesc,
+		"Input machine learning model attributes.")
+	Field(3, "dataset", DatasetArgumentDesc,
+		"Input dataset attributes.")
+	Field(4, "runnable", RunnableArgumentDesc,
+		"Input runnable attributes.")
+})
+
+// RunnableOutputArtifactKind encodes the type of output artifact as well as type-specific attributes
+var RunnableOutputArtifactKind = Type("RunnableOutputArtifactKind", func() {
+	Field(1, "codeset", CodesetArgumentDesc,
+		"Output codeset description.")
+	Field(2, "model", ModelArgumentDesc,
+		"Input machine learning model attributes.")
+	Field(3, "dataset", DatasetArgumentDesc,
+		"Input dataset attributes.")
+	Field(4, "runnable", RunnableArgumentDesc,
+		"Input runnable attributes.")
+})
```

Comment:
It might make sense to join the two, for the sake of reducing code duplication, if not some other reason, but I had a strong feeling that these will diverge and that's why I used two separate data types at the time. In fact, I started by using arrays for inputs but not for outputs, given that the inputs might express _requirements_ that the actual input artifact contents needs to provide, and having a list of variants made sense at the time. 

I will merge them.

Reply:
done